### PR TITLE
feat(Rest): add support for oauth2 endpoints

### PR DESCRIPTION
**Pull request by ckohen**
**Please describe the changes this PR makes and why it should be merged:**

Closes: #90

Alternative to #92 with theoretically significantly better performance on every request and support for the very different error format from oauth

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/rest/src/lib/RequestManager.ts [hunk 1]
```diff
@@ -54,13 +54,19 @@ export interface RequestData {
...
+	body?: BodyInit | unknown;
 	/**
 	 * Additional headers to add to this request
 	 */
 	headers?: Record<string, string>;
+	/**
+	 * Whether to pass-through the body property directly to `fetch()`.
+	 * <warn>This only applies when attachments is NOT present</warn>
+	 */
+	passThroughBody?: boolean;
```
Comment (iCrawl) [inline packages/rest/src/lib/RequestManager.ts:69]: Passthrough is one word.

Comment (ckohen) [reply packages/rest/src/lib/RequestManager.ts:69]: Isn't it both? Where passthrough is used as a noun, e.g. a passthrough panel, and pass through / pass-through is used like this to say the body is allowed to pass through.

I did struggle with capitalization on this, so want to make sure its right though.

Comment (kyranet) [reply packages/rest/src/lib/RequestManager.ts:69]: It is both, as a fact, Node.js has a class named [`PassThrough`](https://nodejs.org/api/stream.html#class-streampassthrough).

**Review by iCrawl — DISMISSED** [review]

**Review by SpaceEEC — APPROVED** [review]

File: packages/rest/src/lib/errors/DiscordAPIError.ts [hunk 2]
```diff
@@ -17,6 +17,11 @@ export interface DiscordErrorData {
 	errors?: DiscordError;
 }
 
+export interface OauthErrorData {
```
Comment (vladfrangu) [inline packages/rest/src/lib/errors/DiscordAPIError.ts:20]: `OAuth`

**Review by vladfrangu — CHANGES_REQUESTED** [review]
How does this actually add support for oauth2 endpoints? You need a bearer token for it, so you'd need to make a second instance of REST for this, is that intended?

**Comment by ckohen** [comment]
Yes, second instance for *utilizing* oauth, which is separate from the literal `/oauth2` endpoints, this is for those paths like /oauth2/token which require a Bot Token

**Comment by vladfrangu** [comment]
`/oauth2/token` doesn't require a bot token... 🤔

**Comment by ckohen** [comment]
Sorry, i meant tokenless 🤦

**Comment by vladfrangu** [comment]
Gotcha. Can you resolve the requested change then? 🙏

**Comment by ckohen** [comment]
yeah, just going through notifs, about to address changes

**Review by vladfrangu — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Pull request merged**